File: tdd_practice/cafekiosk/src/main/java/com/example/cafekiosk/spring/api/controller/order/OrderController.java
package com.example.cafekiosk.spring.api.controller.order;

import com.example.cafekiosk.spring.api.ApiResponse;
import com.example.cafekiosk.spring.api.service.order.OrderService;
import com.example.cafekiosk.spring.api.controller.order.request.OrderCreateRequest;
import com.example.cafekiosk.spring.api.service.order.response.OrderResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.Valid;
import java.time.LocalDateTime;

@RestController
@RequiredArgsConstructor
public class OrderController {

    private final OrderService orderService;

    @PostMapping("/api/v1/orders/new")
    public ApiResponse<OrderResponse> createOrder(@Valid @RequestBody OrderCreateRequest request) {
        return ApiResponse.ok(orderService.createOrder(request.toServiceRequest(), LocalDateTime.now()));
    }
}

File: distributed-transaction/msa-order/src/main/kotlin/org/example/order/controller/OrderController.kt
package org.example.order.controller

import org.example.order.application.OrderCoordinator
import org.example.order.application.OrderService
import org.example.order.application.RedisLockService
import org.example.order.application.dto.CreateOrderResult
import org.example.order.controller.dto.CreateOrderRequest
import org.example.order.controller.dto.PlaceOrderRequest
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class OrderController(
    private val orderService: OrderService,
    private val redisLockService: RedisLockService,
    private val orderCoordinator: OrderCoordinator,
) {

    @PostMapping("/orders")
    fun createOrder(
        @RequestBody request: CreateOrderRequest
    ): CreateOrderResult {
        val result = orderService.createOrder(request.toCommand())
        return CreateOrderResult(result.orderId)
    }

    @PostMapping("/orders/place")
    fun placeOrder(
        @RequestBody request: PlaceOrderRequest
    ) {
        val key = "order:${request.orderId}"

        val acquiredLock = redisLockService.tryLock(key, request.orderId.toString())

        if (!acquiredLock) {
            throw RuntimeException("락 획득 실패")
        }

        try {
            orderCoordinator.placeOrder(request.toCommand())
        } finally {
            redisLockService.releaseLock(key)
        }
    }
}